design/back.compatibility.md

```diff
@@ -0,0 +1,79 @@
+2. move all generation-specific code into a separate repository, and have the main repository include this repository as a submodule for each generation.
+
+* Pros for 1: Easy to use with git. 
+* Cons for 1: It's clunky to have a prefix on filenames (but makes it much harder to accidentally modify things in a final generation - many tools highlight filename but not path). It's still possible to modify the wrong generation's files in the repo.
+* Pros for 2: Makes it harder to do accidental checkins in the wrong repo - you commit into a specific version.
+* Cons for 2: git submodule workflow is clunky, not integrated well with visual git tools. While it seems like meging across generations is easier, it's not something we'll see in practice as fixing finalized generations is not something we should be doing.
+
+I recommend version 1. The pain of the git submodule workflow, and loss of productivity is not worth the minor benefits of 2
+
+### Final Binaries
+Here are two options for building the the binaries
+1. Build a single DLL that includes all generations, with a small bit of generation-shared code for connecting to a server, finding the server version, and choosing the generation-specific mre. 
+2. Build one DLLs per generation, plus a management DLL. 
+Pros for 1: it's easier to have a single output DLL than multiple, as build scripts don't need to be updated when a generation is added. Total binary size is smaller than 2, as compiler will probably optimize out many functions that are unchanged across generations. Con: Higher build time for the DLL.
+I recommend option 1. 
+
+
+### Summary of Recommended Design
+We'll duplicate the files into a different subfolder per generation, add a generation-specific prefix to each file, and leave only the absolutely minimal set of files outside this folder. All of the code will be built into a single DLL.
+
+### Example
+For example, imagine the MRE SDK is at version 2.17.0. It would then support 3 generations of MREs: 1.0 (gen 1), 2.0 (gen 2), and 2.17.0 (pre-release for gen 3). 
+Any host app that includes this plugin will get permanent support for gen 1 and 2, and un-guaranteed prototype of gen3. If the host app later updates to SDK version 2.25.0, then gen 1 and gen 2 MREs run unchanged, but MREs built with gen3 version 2.17 to 2.24 may need to be updated 
+
+### Code to be shared across generations
+The more code is duplicated, the less chance of changes having side effects that spill into finalized generations, so the goal is to move as much as possible to be generational. There is *some* code that must be shared, though, in particular anything that interfaces with the main engine.
+
+#### Shared
+* Initial Connection code. Reason: until host connect to the MRE, it doesn't know the MRE's generation.
+* Code that the host app uses to initialize the MRE system and initialize a single MRE. Reason: host app doesn't need to care about which version each MRE is (MREAppsAPI)
```

Comment:
>Code that the host app uses to initialize the MRE system and initialize a single MRE [](start = 2, length = 84)

I disagree with this one. We've been making regular updates to the plugin interfaces, so we need to be able to finalize those as well. Maybe we have shared code to determine the version of the MRE at a URL, then duplicate the initialization APIs, and let the host initialize only the generation they want.

Reply:
Adding in a third here that the API interface must be versioned in some kind of fashion.

Reply:
I don't think I my intent was clear here, and discussing with Tom highlighted that we were in sync. The main motivation here is that host app code should never have to do "if gen 1 then do A, if gen 2 then do B", but that the host app should just interface with the shared code, and the shared code is then responsible for interfacing with the generational code.

I expect the actual code that connects the host app and the shared code to change (and mostly grow) as new features are added for newer generations. In other words, when a host app imports a newer MRE runtime package, code will not just compile until host app code has been fixed up.